File: lab/example/go/internal/ui/handler.go
// Package ui — HTTP-Handler, Input-Validierung.
// ADR-0001 Layer: darf service + types importieren; KEIN index, embedding direkt.
package ui

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"

	"github.com/example/docsearch/internal/service"
	"github.com/example/docsearch/internal/types"
)

// Handler — HTTP-Layer.
type Handler struct {
	s  *service.Searcher
	ix *service.Indexer
}

// NewHandler konstruiert mit injiziertem Service.
func NewHandler(s *service.Searcher) *Handler {
	return &Handler{s: s}
}

// NewHandlerWithIndexer konstruiert zusätzlich mit dem Indexierungs-Service
// (LH-FA-01).
func NewHandlerWithIndexer(s *service.Searcher, ix *service.Indexer) *Handler {
	return &Handler{s: s, ix: ix}
}

// statusFor bildet die Fehler-Codes aus spec/spezifikation.md §4 auf
// HTTP-Status ab. Unbekannte Fehler sind E099.
func statusFor(err error) (int, string) {
	switch {
	case errors.Is(err, service.ErrDirNotFound):
		return http.StatusBadRequest, "E001"
	case errors.Is(err, service.ErrEmptyQuery):
		return http.StatusBadRequest, "E002"
	case errors.Is(err, service.ErrEmbeddingDown):
		return http.StatusServiceUnavailable, "E003"
	default:
		return http.StatusInternalServerError, "E099"
	}
}

// ReindexHTTP — POST /reindex (LH-FA-01).
func (h *Handler) ReindexHTTP(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Dir string `json:"dir"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		// Unklassifiziert -> E099/500, wie jeder andere unbekannte Fehler
		// (spec/spezifikation.md §4). Eine Zuordnungsstelle, nicht zwei.
		status, code := statusFor(err)
		http.Error(w, `{"error":"`+code+`"}`, status)
		return
	}
	n, err := h.ix.Reindex(req.Dir)
	if err != nil {
		status, code := statusFor(err)
		http.Error(w, `{"error":"`+code+`"}`, status)
		return
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]any{"indexed_docs": n})
}

// SearchHTTP — POST /search.
func (h *Handler) SearchHTTP(w http.ResponseWriter, r *http.Request) {
	var req types.SearchRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		status, code := statusFor(err)
		http.Error(w, `{"error":"`+code+`"}`, status)
		return
	}
	resp, err := h.s.Search(req)
	if err != nil {
		status, code := statusFor(err)
		http.Error(w, `{"error":"`+code+`"}`, status)
		return
	}
	if resp.KClamped {
		w.Header().Set("X-Topk-Clamped", strconv.Itoa(types.MaxTopK))
	}
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]any{"results": resp.Results})
}

File: lab/example/csharp/src/DocSearch/UI/Handler.cs
// UI-Layer. ADR-0001: importiert Service + Types; KEIN Index, Embedding direkt.
using DocSearch.Service;
using DocSearch.Types;

namespace DocSearch.UI
{
    /// <summary>Antwort des UI-Layers: Status und Körper, ohne echten HTTP-Server.</summary>
    public sealed record HttpResult(int Status, IReadOnlyDictionary<string, object> Body);

    /// <summary>Status und Fehler-Code eines abgebildeten Fehlers.</summary>
    public sealed record ErrorMapping(int Status, string Code);

    public sealed class Handler(Searcher searcher, Indexer? indexer = null)
    {
        /// <summary>
        /// Bildet die Fehler-Codes aus spec/spezifikation.md §4 auf HTTP-Status
        /// ab. Unbekannte Fehler sind E099.
        /// </summary>
        public static ErrorMapping StatusFor(Exception exc)
        {
            return exc switch
            {
                DirectoryMissingException => new ErrorMapping(400, "E001"),
                EmptyQueryException => new ErrorMapping(400, "E002"),
                EmbeddingUnavailableException => new ErrorMapping(503, "E003"),
                _ => new ErrorMapping(500, "E099"),
            };
        }

        public SearchResponse HandleSearch(SearchRequest req)
        {
            return searcher.Search(req);
        }

        /// <summary>Suche mit Fehler-Abbildung nach spec/spezifikation.md §4.</summary>
        public HttpResult HandleSearchHttp(SearchRequest req)
        {
            try
            {
                SearchResponse resp = searcher.Search(req);
                return new HttpResult(
                    200, new Dictionary<string, object> { ["results"] = resp.Results });
            }
            catch (Exception e)
            {
                return Failure(e);
            }
        }

        /// <summary>Reindex (LH-FA-01) mit Fehler-Abbildung nach spec §4.</summary>
        public HttpResult HandleReindex(string directory)
        {
            if (indexer is null)
            {
                return new HttpResult(
                    500, new Dictionary<string, object> { ["error"] = "E099" });
            }
            try
            {
                int n = indexer.Reindex(directory);
                return new HttpResult(
                    200, new Dictionary<string, object> { ["indexed_docs"] = n });
            }
            catch (Exception e)
            {
                return Failure(e);
            }
        }

        private static HttpResult Failure(Exception exc)
        {
            ErrorMapping m = StatusFor(exc);
            return new HttpResult(
                m.Status, new Dictionary<string, object> { ["error"] = m.Code });
        }
    }
}
